arroyo/dead_letter_queue/dead_letter_queue.py

```diff
@@ -0,0 +1,54 @@
+import logging
+from typing import Optional
+from arroyo.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessage,
+)
+from arroyo.processing.strategies.abstract import ProcessingStrategy as ProcessingStep
+from arroyo.types import Message, TPayload
+
+logger = logging.getLogger(__name__)
+
+
+class DeadLetterQueue(ProcessingStep[TPayload]):
+    """
+    DLQ Processing Step.
+
+    Attempts to submit a given message to the next processing step,
+    handling an `InvalidMessage` according to the given Policy.
+    """
+
+    def __init__(
+        self,
+        next_step: ProcessingStep[TPayload],
+        policy: DeadLetterQueuePolicy[TPayload],
+    ) -> None:
+        self.__next_step = next_step
+        self.__policy = policy
+        self.__closed = False
+
+    def poll(self) -> None:
+        self.__next_step.poll()
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+        try:
+            self.__next_step.submit(message)
+        except InvalidMessage as e:
+            logger.warning(
+                f"Invalid Message caught by Dead Letter Queue: {message}",
+                exc_info=True,
+            )
```

Comment:
We cannot have a warning every time. Let the policy decide whether to log. 
If the dead letter queue ends up receiving InvalidMessage exceptions for all messages on a high volume topic we would be floodding ourselves with messages to log (they all go in sentry).
So Arroyo does not have enough context to decide if an invalid message should be logged or not. the policy can have such context.
Recording a metric is a better option.